include/picongpu/particles/collision/IntraCollision.hpp

```diff
@@ -0,0 +1,232 @@
+                    uint32_t numParticlesInSupercell = superCell.getNumParticles();
+
+                    auto accFilter = filter(acc, localSuperCellOffset, threads::WorkerCfg<T_numWorkers>{workerIdx});
+
+                    /* loop over all particles in the frame */
+                    ForEachIdx<FrameDomCfg> forEachFrameElem(workerIdx);
+                    FramePtr firstFrame = pb.getFirstFrame(superCellIdx);
+
+                    prepareList(
+                        acc,
+                        forEachFrameElem,
+                        deviceHeapHandle,
+                        pb,
+                        firstFrame,
+                        numParticlesInSupercell,
+                        parCellList,
+                        nppc,
+                        accFilter);
+
+
+                    cellDensity(acc, forEachFrameElem, firstFrame, pb, parCellList, densityArray, accFilter);
+                    cupla::__syncthreads(acc);
+
+                    // shuffle  indices list
+                    forEachFrameElem([&](uint32_t const linearIdx, uint32_t const) {
+                        parCellList[linearIdx].shuffle(acc, rngHandle);
+                    });
+
+                    memory::CtxArray<
+                        decltype(collisionFunctor(
+                            acc,
+                            alpaka::core::declval<DataSpace<simDim> const>(),
+                            /*frameSize is used because each virtual worker
+                             * is creating **exactly one** functor
+                             */
+                            alpaka::core::declval<WorkerCfg<frameSize> const>(),
+                            alpaka::core::declval<float_X const>(),
+                            alpaka::core::declval<float_X const>(),
+                            alpaka::core::declval<uint32_t const>(),
+                            alpaka::core::declval<float_X const>())),
+                        FrameDomCfg>
+                        collisionFunctorCtx{};
+
+                    forEachFrameElem([&](uint32_t const linearIdx, uint32_t const idx) {
+                        uint32_t const sizeAll = parCellList[linearIdx].size;
+                        if(sizeAll < 2u)
+                            return;
+                        // skip particle offset counter
+                        uint32_t* listAll = parCellList[linearIdx].ptrToIndicies;
+                        uint32_t potentialPartners = sizeAll - 1u + sizeAll % 2u;
+                        collisionFunctorCtx[idx] = collisionFunctor(
+                            acc,
+                            localSuperCellOffset,
+                            WorkerCfg<T_numWorkers>{workerIdx},
+                            densityArray[linearIdx],
+                            densityArray[linearIdx],
+                            potentialPartners,
+                            coulombLog);
+                        for(uint32_t i = 0; i < sizeAll; i += 2)
```

Comment:
BUG??? should be `for(uint32_t i = 0; i + 2 <= sizeAll; i += 2)` else the last particle pare could be outside of memory.
Same happened in InterCollision.

Reply:
` auto parOdd = detail::getParticle(pb, firstFrame, listAll[iOdd * 2 + 1]);`
Will access particle `i+1`

Reply:
IN my original code this was: https://github.com/psychocoderHPC/picongpu/blob/48359030e9ea8cc34038cb9da3a3a9f25ba382da/include/picongpu/particles/collision/IntraSpecies.hpp#L359

Maybe your change work because of the next line 
```
uint32_t const sizeOdd = sizeAll / 2;
uint32_t const iOdd = (i / 2) % sizeOdd;
```
but from pure looking to the loop it is hard to see if it is safe.

Reply:
I simplified the loop significantly. Now it looks more like what you did.